_internal/utils/mutate.ts

```diff
@@ -4,23 +4,40 @@ import { SWRGlobalState } from './global-state'
 import { getTimestamp } from './timestamp'
 import * as revalidateEvents from '../constants'
 import {
-  Key,
   Cache,
   MutatorCallback,
   MutatorOptions,
   GlobalState,
-  State
+  State,
+  Falsy,
+  Arguments
 } from '../types'
 
+type KeyFilter = (key?: string) => boolean
+
 export const internalMutate = async <Data>(
   ...args: [
     Cache,
-    Key,
+    KeyFilter | Arguments | Falsy,
     undefined | Data | Promise<Data | undefined> | MutatorCallback<Data>,
     undefined | boolean | MutatorOptions<Data>
   ]
-): Promise<Data | undefined> => {
-  const [cache, _key, _data, _opts] = args
+): Promise<(Data | undefined)[]> => {
```

Comment:
We shouldn't break the existing use case where `mutate` returns just the bare `Data`. We should overload the function type to only return `Data[]` if a filter function is passed in.